File: libgomp/testsuite/libgomp.fortran/target-has-device-addr-1.f90
program main
  use omp_lib
  use iso_c_binding
  implicit none

  integer, parameter :: N = 40
  integer :: x, i
  integer :: y (N)
  integer :: u (0)

  x = 24
  !$omp target data map(x) use_device_addr(x)
    !$omp target has_device_addr(x)
      x = 42;
    !$omp end target
  !$omp end target data
  if (x /= 42) stop 1

  y = 42
  !$omp target data map(y) use_device_addr(y)
    !$omp target has_device_addr(y)
      y = [(i, i=1, N)]
    !$omp end target
  !$omp end target data
  if (any (y /= [(i, i = 1, N)])) stop 2

  !$omp target data map(y(:N)) use_device_addr(y)
    !$omp target has_device_addr(y(:N))
      y = [(i+2, i=1, N)]
    !$omp end target
  !$omp end target data
  if (any (y /= [(i+2, i = 1, N)])) stop 3

  !$omp target data map(y) use_device_addr(y)
    !$omp target has_device_addr(y(24:))
      do i = 24, N
        y(i) = i + 3
      end do
    !$omp end target
  !$omp end target data
  do i = 24, N
    if (y(i) /= i + 3) stop 5
  end do

  !$omp target data map(u) use_device_addr(u)
    !$omp target has_device_addr(u)
    !$omp end target
  !$omp end target data

end program main

File: libgomp/testsuite/libgomp.c-c++-common/target-has-device-addr-1.c
/* Testing the 'has_device_addr' clause on the target construct with
   enclosing 'target data' construct. */

#define N 40

int
main ()
{
  int x = 24;

  #pragma omp target data map(x) use_device_addr(x)
    #pragma omp target has_device_addr(x)
      x = 42;
  if (x != 42)
    __builtin_abort ();

  int y[N];

  for (int i = 0; i < N; i++)
    y[i] = 42;
  #pragma omp target data map(y) use_device_addr(y)
    #pragma omp target has_device_addr(y)
      for (int i = 0; i < N; i++)
	y[i] = i;
  for (int i = 0; i < N; i++)
    if (y[i] != i)
      __builtin_abort ();

  #pragma omp target data map(y[:N]) use_device_addr(y)
    #pragma omp target has_device_addr(y[:N])
      for (int i = 0; i < N; i++)
	y[i] = i + 2;
  for (int i = 0; i < N; i++)
    if (y[i] != i + 2)
      __builtin_abort ();

  #pragma omp target data map(y[:N]) use_device_addr(y)
    #pragma omp target has_device_addr(y[24])
	y[24] = 42;
  if (y[24] != 42)
    __builtin_abort ();

  #pragma omp target data map(y[:N]) use_device_addr(y)
    #pragma omp target has_device_addr(y[24:])
      for (int i = 24; i < N; i++)
	y[i] = i + 3;
  for (int i = 24; i < N; i++)
    if (y[i] != i + 3)
      __builtin_abort ();

  #pragma omp target data map(y[:N]) use_device_addr(y)
    #pragma omp target has_device_addr(y[12:24])
      for (int i = 12; i < 24; i++)
	y[i] = i + 4;
  for (int i = 12; i < 24; i++)
    if (y[i] != i + 4)
      __builtin_abort ();

  int u[0];
  #pragma omp target data map(u) use_device_addr(u)
    #pragma omp target has_device_addr(u)
  ;

  struct S { int m; } s;
  s.m = 42;
  #pragma omp target data map (s) use_device_addr (s)
    #pragma omp target has_device_addr (s)
      ++s.m;
  if (s.m != 43)
    __builtin_abort ();

  return 0;
}
